go/runtime/bundle/discovery.go

```diff
@@ -0,0 +1,452 @@
+		case <-d.discoverCh:
+		case <-ctx.Done():
+			d.logger.Info("stopping discovery")
+			return
+		}
+
+		_ = d.Discover()
+		d.Download()
+	}
+}
+
+// Discover searches for new bundles in the bundle directory and adds them
+// to the bundle registry.
+func (d *Discovery) Discover() error {
+	d.logger.Debug("discovering bundles")
+
+	entries, err := os.ReadDir(d.bundleDir)
+	if err != nil {
+		d.logger.Error("failed to read bundle directory",
+			"err", err,
+			"dir", d.bundleDir,
+		)
+		return fmt.Errorf("failed to read bundle directory: %w", err)
+	}
+
+	for _, entry := range entries {
+		filename := entry.Name()
+		if entry.IsDir() || filepath.Ext(filename) != FileExtension {
+			continue
+		}
+
+		baseFilename := strings.TrimSuffix(filename, FileExtension)
+		if len(baseFilename) != 2*hash.Size {
+			continue
+		}
+
+		var manifestHash hash.Hash
+		if err = manifestHash.UnmarshalHex(baseFilename); err != nil {
+			continue
+		}
+
+		if d.registry.HasBundle(manifestHash) {
+			continue
+		}
+
+		d.logger.Info("found new bundle",
+			"file", filename,
+		)
+
+		path := filepath.Join(d.bundleDir, filename)
+		if err = d.registry.AddBundle(path, manifestHash); err != nil {
+			d.logger.Error("failed to add bundle to registry",
+				"err", err,
+				"path", path,
+			)
+			return fmt.Errorf("failed to add bundle to registry: %w", err)
+		}
+	}
+
+	return nil
+}
+
+// Queue updates the checksums of bundles that need to be downloaded
+// for the given runtime.
+func (d *Discovery) Queue(runtimeID common.Namespace, manifestHashes []hash.Hash) {
+	d.mu.Lock()
+	defer d.mu.Unlock()
+
+	// Download bundles only if at least one endpoint is configured.
+	if len(d.globalBaseURLs) == 0 && len(d.runtimeBaseURLs[runtimeID]) == 0 {
+		return
+	}
+
+	// Filter out bundles that have already been fetched.
+	var hashes []hash.Hash
+	for _, hash := range manifestHashes {
+		if d.registry.HasBundle(hash) {
+			continue
+		}
+		hashes = append(hashes, hash)
+	}
+
+	// Update the queue with the new hashes.
+	if len(hashes) == 0 {
+		delete(d.manifestHashes, runtimeID)
+		return
+	}
+	d.manifestHashes[runtimeID] = hashes
+
+	// Trigger immediate discovery or download of new bundles.
+	select {
+	case d.discoverCh <- struct{}{}:
+	default:
+	}
+}
+
+// Download tries to download bundles in the queue.
+func (d *Discovery) Download() {
+	d.mu.RLock()
+	runtimeIDs := slices.Collect(maps.Keys(d.manifestHashes))
+	d.mu.RUnlock()
+
+	for _, runtimeID := range runtimeIDs {
+		d.downloadBundles(runtimeID)
+	}
+}
+
+func (d *Discovery) downloadBundles(runtimeID common.Namespace) {
+	// Try to download queued bundles.
+	d.mu.RLock()
+	hashes := d.manifestHashes[runtimeID]
+	d.mu.RUnlock()
+
+	downloaded := make(map[hash.Hash]struct{})
+	for _, hash := range hashes {
+		if err := d.downloadBundle(runtimeID, hash); err != nil {
+			d.logger.Error("failed to download bundle",
+				"err", err,
+				"runtime_id", runtimeID,
+				"manifest_hash", hash.Hex(),
+			)
```

Comment:
Looking at `d.downloadBundle` we join the errors from all the downloads, meaning unless the first `baseUrl` is success, we would receive error here even if we successfully download the bundle using e.g. third baseUrl.

I know we plan to start with single url for the start so this should not be problematic for now, but might become misleading in the future?

Reply:
No. Observe that we only return error if all downloads fail. So if the first two fail and the third one doesn't, the `downloadBundle` will return `nil`.

Reply:
Ah my bad. sorry!